### PR TITLE
Fix computing averages in WindowOperatorStats

### PR DESCRIPTION
**Pull request by losipiuk**
The code failed if all the `WindowInfo` objects we iterated over had `getTotalRowsCount() == 0`.
This change fixes the behaviour but I am not exactly sure why we need the `.filter(windowInfo -> windowInfo.getTotalRowsCount() > 0)` alltogether.
@arhimondr can you take a look if this is a proper fix and if original code is right?

cc: @findepi

### CODE REVIEW

**Comment by findepi** [comment]
It would be good to have some test. However, I don't know how to reproduce the original problem in a synthetic way (without a cluster).

**Comment by arhimondr** [comment]
@losipiuk I'm not sure i can remember this code. Maybe @maciejgrzybek would've a better input here.

The fix you introduced seems to be reasonable. Can't you reproduce the issue by applying window function to the empty data set or "sparse" data set with empty partitions/batches?

**Comment by losipiuk** [comment]
> Can't you reproduce the issue by applying window function to the empty data set or "sparse" data set with empty partitions/batches?

I will try to add test.

The change fixes the issue. What I was more asking is why we are filtering the windows without rows in the first place.
@maciejgrzybek ?

**Comment by losipiuk** [comment]
Added test

File: presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestDistributedQueries.java [hunk 1]
```diff
@@ -289,6 +289,12 @@ public void testExplainAnalyze()
         assertExplainAnalyze("EXPLAIN ANALYZE SHOW SESSION");
     }
 
+    @Test
+    public void testExplainAnalyzeVerbose()
+    {
+        assertExplainAnalyze("EXPLAIN ANALYZE VERBOSE SELECT * FROM orders");
```
Comment (findepi) [inline presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestDistributedQueries.java]: `assertExplainAnalyze` should be called `assertDoesn’tFail` ...

File: presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestDistributedQueries.java [hunk 2]
```diff
@@ -289,6 +289,12 @@ public void testExplainAnalyze()
         assertExplainAnalyze("EXPLAIN ANALYZE SHOW SESSION");
     }
 
+    @Test
+    public void testExplainAnalyzeVerbose()
```
Comment (findepi) [inline presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestDistributedQueries.java]: - Is there a point in executing this for many different connectors?

- Commits are out of order.

File: presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestDistributedQueries.java [hunk 3]
```diff
@@ -293,6 +293,8 @@ public void testExplainAnalyze()
     public void testExplainAnalyzeVerbose()
     {
         assertExplainAnalyze("EXPLAIN ANALYZE VERBOSE SELECT * FROM orders");
+        assertExplainAnalyze("EXPLAIN ANALYZE VERBOSE SELECT rank() OVER (PARTITION BY orderkey ORDER BY clerk DESC) FROM orders");
+        assertExplainAnalyze("EXPLAIN ANALYZE VERBOSE SELECT rank() OVER (PARTITION BY orderkey ORDER BY clerk DESC) FROM orders WHERE orderkey < 0");
```
Comment (findepi) [inline presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestDistributedQueries.java]: Is there a point in executing these for many different connectors?

Comment (losipiuk) [reply presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestDistributedQueries.java]: Concerning this specific issue - no I don't think so.
Yet generally the explain output can be different for different connectors, and potentially we can have connector dependent failures here.
What do you propose?

Comment (findepi) [reply presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestDistributedQueries.java]: I think it's good to have _some_ run of `EXPLAIN ANALYZE VERBOSE` for all connectors. And this should include single-table selects and eg a  join.

For the actual problem, I would propose introduction of `TestExplain` test class. (Just like `TestGrouping` but with distributed query runner)

Comment (losipiuk) [reply presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestDistributedQueries.java]: Not that easy as it looks like. It would make sense to migrate the most of explain tests from `ATQ` to that new class. Yet that is not trivial because the machinery for testing explains lives in  `AbstractTestQueryFramework` superclass. This refactor is too much of an effort IMO for this issue. I will leave it out.

**Comment by maciejgrzybek** [comment]
IIRC, the idea was to distinguish two type of problems:
1. Data skew (should be visible in `ExchangeOperator` stats), e.g. when one partition is significantly larger than others. In such case, WindowOperator will be inefficient because parallelism will go down.
2. Problems in WindowOperator execution itself.

If you included empty indexes in WO stats, the active WO stats would be overwhelmed by data skew problems.
Imagine situation when all data is in one partition (processed by 8 local WOs) but you have 1024 WOs total in plan (128 nodes are idling till 1 node finishes processing WO).
In such case you care how efficiently the only occupied 8 local operators were working (how many times index was rebuilt, how big it was etc, did everyone get the same index or maybe there was some skew which could be prevent on the operator level etc.).
If you included all 127 remaining idling nodes in the average, the problems with running operators would be not visible (in most cases you'd see the average va lues very low, no matter how well active operators are doing).

So, in other words, it's to make stats more granular:
if you're looking for data distribution skew, keep an eye on `ExchangeOperator` stats.
If you're debugging execution of WO, focus on relevant stats for WOs.

Keep in mind, though, that WO stats are very low level and tightly coupled to implementation. They don't make sense to DBA, if he doesn't know execution engine and implementation of WOs.
It's mostly to help driving window function improvements by better understanding why some queries are slower than others.

Does it make sense?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/WindowOperatorStats.java [hunk 4]
```diff
@@ -53,13 +53,13 @@ public static WindowOperatorStats create(WindowInfo info)
...
                 .average()
-                .getAsDouble();
+                .orElse(Double.NaN);
 
         double averageNumberOfRows = info.getWindowInfos().stream()
                 .filter(windowInfo -> windowInfo.getTotalRowsCount() > 0)
                 .mapToLong(DriverWindowInfo::getTotalRowsCount)
                 .average()
-                .getAsDouble();
+                .orElse(Double.NaN);
```
Comment (maciejgrzybek) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/WindowOperatorStats.java]: In line 80 (https://github.com/prestodb/presto/pull/10200/commits/8b643f81bc4ee6504c8f768fc5f0bf5ffea01cef#diff-cf5d7dbb1d37bd1638bd50937d1152f0R80) we're using this value in subtraction. X - NaN = NaN. This is not what we want to report. Please handle that case.

**Comment by losipiuk** [comment]
> Does it make sense?

Thanks for explanation @maciejgrzybek . That makes some sense (in terms there are cases when it helps to debug things, otherwise hard to tackle). I will leave it.
Added commit which adds your explanation to the code base in comment.
